src/brogue/IO.c

```diff
@@ -3168,19 +3218,42 @@ void messageWithoutCaps(char *msg, enum messageFlags flags) {
         }
     }*/
 
-    for (i = MESSAGE_LINES - 1; i >= 1; i--) {
-        messageConfirmed[i] = messageConfirmed[i-1];
-        strcpy(displayedMessage[i], displayedMessage[i-1]);
+    // Add the message to the archive, bumping counts for recent duplicates
+    newMessage = true;
+
+    // For each at most MESSAGE_ARCHIVE_ENTRIES - 1 past entries..
+    for (i = 1; i < MESSAGE_ARCHIVE_ENTRIES; i++) {
+        archiveEntry = getArchivedMessage(i);
+
+        // Always consider the latest entry for collapse, but also past ones that arrived this turn
+        if (i > 1 && archiveEntry->turn != rogue.playerTurnNumber) {
```

Comment:
Commit message not withstanding, this iteration of the logic is still here because we don't yet have the flags to distinguish combat messages from non-combat ones.  That arrives in the final commit, which then amends this line.